### PR TITLE
#14 Added Form Field Grouping

### PR DESCRIPTION
**Pull request by qoda**
Added new hierarchy to accommodate field groups in forms.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -34,15 +34,25 @@ FORMFACTORY["field-types"]
...
+``django-formfactory`` provide a base ``FormView`` and ``FactoryWizardView``
+which can both be used directly or subclassed if you require extra context
+or form data processing.
+
+Templates
+~~~~~~~~~
+
+``django-formfactory`` allows you to either override the template for all forms
+by adding a template ``formfactory/form_detail.html`` or an individual form by
+adding a template ``formfactory/form_detail_<form-slug>.html``to your project's
```
Comment (phalamag) [inline README.rst]: Space after ``formfactory/form_detail_<form-slug>.html``

File: formfactory/admin.py [hunk 2]
```diff
@@ -63,13 +62,30 @@ class WizardActionThroughInline(admin.StackedInline):
...
+    inlines = [WizardFormThroughInline, WizardActionThroughInline]
     prepopulated_fields = {"slug": ["title"]}
 
 
+class FieldGroupThroughInline(admin.StackedInline):
+    form = forms.FieldGroupThroughAdminForm
+    model = models.FieldGroupThrough
+
+
+class FormFieldGroupAdmin(admin.ModelAdmin):
```
Comment (phalamag) [inline formfactory/admin.py]: How about we add `prepopulated_fields = {"slug": ["title"]}` to this `ModelAdmin`?

File: formfactory/factory.py [hunk 3]
```diff
@@ -24,55 +29,165 @@ def __init__(self, *args, **kwargs):
         )
 
         # Interates over the fields defined in the Form model and sets the
```
Comment (phalamag) [inline formfactory/factory.py]: *Iterates

File: formfactory/factory.py [hunk 4]
```diff
@@ -24,55 +29,165 @@ def __init__(self, *args, **kwargs):
...
-                label=field.label,
-                initial=field.initial,
-                required=field.required,
-                disabled=field.disabled,
-                help_text=field.help_text,
-                validators=additional_validators
+        # appropriate attributes and builds up the fieldgroups.
+        self.field_group = []
+        for field_group in defined_field_groups:
+            fields = field_group.fields.all()
```
Comment (phalamag) [inline formfactory/factory.py]: I think you're missing `.order_by("fieldgroupthrough")`

File: formfactory/models.py [hunk 5]
```diff
@@ -140,7 +140,7 @@ def as_form(self, data=None, files=None):
             )
 
         return factory.FormFactory(
-            data, files, prefix=self.slug, fields=self.fields.all(),
+            data, files, prefix=self.slug, field_groups=self.fieldgroups.all(),
```
Comment (phalamag) [inline formfactory/models.py]: The `self.fieldgroups.all()` should probably be ordered by `"fieldgroupformthrough"`

File: formfactory/models.py [hunk 6]
```diff
@@ -204,18 +204,48 @@ def __unicode__(self):
...
+class FieldGroupFormThrough(models.Model):
+    """Through table for field groups forms with a defined order.
+    """
+    field_group = models.ForeignKey(FormFieldGroup)
+    form = models.ForeignKey(Form)
+    order = models.PositiveIntegerField(default=0)
+
+    class Meta(object):
+        ordering = ["order"]
+        verbose_name = "Form"
```
Comment (phalamag) [inline formfactory/models.py]: `verbose_name` should be `"Field group"`

File: formfactory/models.py [hunk 7]
```diff
@@ -204,18 +204,48 @@ def __unicode__(self):
...
+        verbose_name_plural = "Forms"
```
Comment (phalamag) [inline formfactory/models.py]: `verbose_name_plural = "Field groups"`

File: formfactory/models.py [hunk 8]
```diff
@@ -235,8 +265,21 @@ class FormField(models.Model):
...
+    """Through table for form fields and field groups with a defined order.
+    """
+    field = models.ForeignKey(FormField)
+    field_group = models.ForeignKey(FormFieldGroup)
+    order = models.PositiveIntegerField(default=0)
+
     class Meta(object):
-        ordering = ["position"]
+        ordering = ["order"]
+        verbose_name = "Field Group"
```
Comment (phalamag) [inline formfactory/models.py]: `verbose_name` should be `"Field"`

File: formfactory/admin.py (hunk 2 above)
Comment (qoda) [inline formfactory/admin.py]: ModelAdmin is inherited by everything and is django native, so would require a monkey patch which would affect all classes inheriting from ModelAdmin. A BaseModelAdmin class could be added but it's not necessary for such a small set of admin classes which require the prepopulate fields var.

Comment (phalamag) [inline formfactory/admin.py]: :+1:

**Comment by phalamag** [comment]
+1

**Pull request merged**